### DISCUSSION TITLE
jsonable_encoder is not async, can stall the event loop

### DISCUSSION THREAD

**Discussion by sm-Fifteen** [fastapi/fastapi#9044] [Questions]
### Describe the bug

I was investigating 502 errors on an API server I'm developping using FastAPI. Basically, the server would intermittently stop responding for such long periods of time that the Apache reverse proxy in front of it would consider it offline and return a Bad Gateway error. I know this is usually a symptom that the event loop is stalled, so I looked through my own code to see if there was any point where I got the blocking/non-blocking routes wrong, and didn't really find much. However, when I looked at the profiler results (which reminds me I still need to get my profiler middleware out there), this stood out:

(duration in milliseconds)
> db_exec;dur=27717.296,
> db_fetchone;dur=182.606,
> 1deps;dur=0.516,
> 3valid;dur=0.041,
> **4encode;dur=6267.253**

That last one, `4encode`, is mapped to `jsonable_encoder`, which is a CPU-bound function run synchronously inside `serialize_response` (an async function). There is an assumption that `jsonable_encoder` (and/or whatever JSON renderer function is used) will never end up running for any significant amount of time, but if it does, the entire server will become unresponsive.

### To Reproduce

While the use case which causes me to encounter that bug has to do with encoding very large sets of complex data, I have a simpler use case which exploits the blocking nature of jsonable encoder:

1. Create a file with:

```py
from fastapi import FastAPI
from time import sleep

def blocking_iterator():
    for num in range(10):
        sleep(1)
        yield num

app = FastAPI()

@app.get("/slow")
def blocks_on_encode():
    # Serialized into a list, will block jsonable_encoder
    # for 10 seconds and stall the whole event loop.
    return blocking_iterator()
```

2. Open two browser tabs and call the endpoint `/slow` **twice**.
3. The first tab returns after 10 seconds, while the second tab only returns 10 seconds later.
4. Additionally, the server will not respond to any attempts to close it until the request it was busy encoding returns a response.

### Expected behavior

JSON encoding and rendering blocking the server is something I would consider highly undesirable for an async-focused framework. With that said, given how Python's GIL works, I don't know if jsonable_encoder can simply be moved into a thread, since jsonable_encoder would then simply block the GIL instead of blocking the event loop, which would end up being no different.

JSON rendering could end up facing similar issues, though I have not yet tested them to see how long they may end up blocking the event loop and/or thread. None of the one I know of would work with threads:
* The [`json` module in the standard library](https://github.com/python/cpython/blob/master/Lib/json/encoder.py) has [an accelerator module in C](https://github.com/python/cpython/blob/master/Modules/_json.c), though it does not release the GIL
* ujson seemingly does not release the GIL, as evidenced by ultrajson/ultrajson#199
* orjson doesn't seem to release the GIL anywhere either, though a handful of functions run with `pyo3::Python::assume_gil_acquired()` (which immediately panics if the current thread does not hold the GIL).

----

This is rather tricky problem, I would say, and I personally can't think of a good solution for it at the moment given Python's limitations. At the very least, the documentation should acknowledge the problem and make it clear that encoding may not scale well for *extremely large* amounts of data, though I don't know if any of the other Python web frameworks fare much better in that regard, and if encoding volumes of data large enough to cause this issue to occur isn't actually considered bad practice in other frameworks regardless of the language.

EDIT: I actually forgot that [the GIL still does *some* scheduling](https://www.dabeaz.com/python/UnderstandingGIL.pdf), so moving serialization to a separate thread might work to prevent computing tasks from blocking the server, even though it could still end up starving the main thread.

### Environment

- OS: Windows 10
- FastAPI Version: 0.52.0
- Python version: 3.8.2

**Comment by sm-Fifteen** [comment]
For what it's worth, a lot of the serialization functionality of `jsonable_encoder` [can be performed by `orjson` directly](https://github.com/ijl/orjson#serialize):

> It natively serializes
`str`, `dict`, `list`, `tuple`, `int`, `float`, `bool`,
`dataclasses.dataclass`, `typing.TypedDict`, `datetime.datetime`,
`datetime.date`, `datetime.time`, `uuid.UUID`, `numpy.ndarray`, and
`None` instances. It supports arbitrary types through `default`.
It does not serialize subclasses of
supported types natively, with the exception of `dataclasses.dataclass`
subclasses.

There might be some solid performance gains to be had on this critical path by allowing the serialization and the rendering steps of this process to be integrated together in some fashion, depending on the response class being used.

**Comment by connebs** [comment]
@sm-Fifteen I'm actually doing that in some of my endpoints that need extra perf.

I just take my python dicts and dump them into an `ORJSONResponse` directly, which avoids `jsonable_encoder`.

Previously I had a situation where one endpoint in particular was naively doing this, because all I was doing was getting my ORM object and then setting a `response_model`:

1. Fetch data from db 
2. Dump into SQLAlchemy ORM
3. Make into Pydantic model
4. Run jsonable_encoder on it
5. ORJSONResponse

Steps 3-5 are just what FastAPI does naturally (if using `response_model`), and is generally fine for 90%+ of payloads which are either relatively small or relatively flat. But this is my main data retrieval endpoint that is frequently serving multi-MB of multi-level data.

Now for that one critical path I'm:

1. Fetching rows async-style and without an ORM using https://github.com/encode/databases
2. Manually turning the returned `Records` into `dicts`
3. Shoving that in an `ORJSONResponse`

Previously I was averaging about 800ms (with a max of like 5s for the very largest payloads) for this endpoint, now I'm getting about 100ms. `jsonable_encoder` was chewing up about 4x the new baseline (400ms), while SQLAlchemy was chewing up about 3x (300ms).

**Comment by salvisolamartinell** [comment]
I don't know if it is related, but this fails:
```python
from unittest.mock import AsyncMock
from fastapi.encoders import jsonable_encoder
jsonable_encoder(AsyncMock())
```
`Fatal Python error: Cannot recover from stack overflow.`

**Comment by SyntaxColoring** [comment]
@tiangolo Thanks for the issue cleanup, but I think this particular thread is more of a bug report on library design than a discussion. Can this be moved back to an issue?

**Reply by sm-Fifteen** [reply]
Discussions also lose cross-issue references, so we can't see how #1359 (now #9059) also relates to this, and included benchmarks and proper preformance analysis.
